spec.html

```diff
@@ -35418,16 +35814,25 @@ <h1>
         </emu-alg>
         <emu-grammar>Atom :: `.`</emu-grammar>
         <emu-alg>
-          1. Let _A_ be the CharSet of all characters.
+          1. Let _A_ be ! GetAllCharacters().
           1. If _DotAll_ is not *true*, then
             1. Remove from _A_ all characters corresponding to a code point on the right-hand side of the |LineTerminator| production.
           1. Return ! CharacterSetMatcher(_A_, *false*, _direction_).
         </emu-alg>
         <emu-grammar>Atom :: CharacterClass</emu-grammar>
         <emu-alg>
           1. Let _cc_ be CompileCharacterClass of |CharacterClass|.
-          1. Return ! CharacterSetMatcher(_cc_.[[CharSet]], _cc_.[[Invert]], _direction_).
+          1. If _UnicodeSets_ is *false* or if every element of _cc_.[[CharSet]] consists of a single character, return ! CharacterSetMatcher(_cc_.[[CharSet]], _cc_.[[Invert]], _direction_).
+          1. Assert: _cc_.[[Invert]] is *false*.
+          1. Let _D_ be the Disjunction containing:
```

Comment:
@bakkot Re: your comment on this. The intent is to logically replace the character class with a disjunction in which the alternatives are sorted longest to shortest, as described here: https://github.com/tc39/proposal-regexp-set-notation#whats-the-match-order-for-character-classes-containing-strings We understand that the current spec text is not correct, and would appreciate help in making it work. @gibson042 had two possible approaches in mind that might work.

Reply:
I believe the right approach is to construct and return a Matcher which loops over the strings in order and attempts to match each of them one character at a time. Fortunately this operation is already returning a Matcher, so you shouldn't need to do any complex rewiring.

You could do this by manually copying the logic in `CompileSubpattern` of `Disjunction :: Alternative | Disjunction` (for the outer loop) plus the logic in `CompileSubpattern` of ` Alternative :: Alternative Term` (for iterating the characters within a string) plus the logic in `CharacterSetMatcher` (for matching each character). That approach would end up duplicating a lot of logic, though.

So instead, I think the best approach would be to refactor the first two things so that they are each defined in terms of a new AO which takes two Matchers and produces a new Matcher, and then these new Matchers can be called in both places that logic is needed.

That is, I suggest defining a `MatchAlternatives` (or some other name) AO, along the lines of

```
The abstract operation MatchAlternatives takes arguments m1 (a Matcher) and m2 (a Matcher). It performs the following steps when called:

1. Return a new Matcher with parameters (x, c) that captures m1 and m2 and performs the following steps when called:
    1. Assert: x is a State.
    1. Assert: c is a Continuation.
    1. Let r be m1(x, c).
    1. If r is not failure, return r.
    1. Return m2(x, c).
```

and then you can redefine `CompileSubpattern` for ` Disjunction :: Alternative | Disjunction` in terms of this:

```
 Disjunction :: Alternative | Disjunction

1. Let m1 be CompileSubpattern of Alternative with argument direction.
1. Let m2 be CompileSubpattern of Disjunction with argument direction.
1. Return MatchAlternatives(m1, m2).
```


And then you can make use of this when constructing the Matcher for a set containing strings: you construct a Matcher for each string in turn and reduce them into a single Matcher by invoking MatchAlternatives repeatedly. (Or if you prefer to think in terms of lists, you could modify this so that MatchAlternatives took a list, and then in `CompileSubpattern` for ` Disjunction :: Alternative | Disjunction` invoke it with a list of length two.)

And then a similar refactoring for `Alternative :: Alternative Term`, by introducing a new `MatchSequence` AO (again, name TBD), which would let you construct the Matcher for an individual string.

Does that make sense?

Reply:
Not forgotten. This makes sense, but I need a little time to try to get this right, so I saved it for last.

Reply:
Please review https://github.com/mathiasbynens/ecma262/pull/4

Reply:
Resolving given the LGTM on the PR.

Reply:
Merged. I know there are editorial issues, but Mathias and Kevin approved it for its functionality.